### PR TITLE
Feature/repay

### PR DESCRIPTION
**Pull request by dariacherniak**

### CODE REVIEW

**Comment by Pavel-Blaize** [comment]
Task linked: [CU-1y7wzwu Repay feature](https://app.clickup.com/t/1y7wzwu)

File: contracts/controller/src/repay.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
+use crate::*;
+
+#[near_bindgen]
+impl Contract {
+
```
Comment (markblaize) [inline contracts/controller/src/repay.rs:5]: @dariacherniak extra line

File: contracts/controller/src/repay.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        true
+    }
+
+    pub fn repay_borrows(
+        &mut self,
+        account_id: AccountId,
+        token_address: AccountId,
+        tokens_amount: WBalance,
+    )-> Balance{
+
```
Comment (markblaize) [inline contracts/controller/src/repay.rs:21]: @dariacherniak extra line

File: contracts/dtoken/src/borrow.rs [hunk 3]
```diff
@@ -1,3 +1,46 @@
 use crate::*;
 
-impl Contract {}
+#[near_bindgen]
+impl Contract {
+
```
Comment (markblaize) [inline contracts/dtoken/src/borrow.rs:5]: @dariacherniak extra line

File: contracts/dtoken/src/borrow.rs [hunk 4]
```diff
@@ -1,3 +1,46 @@
...
+    pub fn decrease_borrows(
+        &mut self,
+        account: AccountId,
+        tokens_amount: WBalance,
+    ) -> Balance {
+        let existing_borrows: Balance = self.get_borrows_by_account(account.clone());
+
+        assert!(existing_borrows >= Balance::from(tokens_amount), "Too much borrowed assets trying to pay out");
```
Comment (markblaize) [inline contracts/dtoken/src/borrow.rs:13]: @dariacherniak user is allowed to repay whole amount so this check is to strict

File: contracts/dtoken/src/borrow.rs [hunk 5]
```diff
@@ -1,3 +1,46 @@
...
+
+    pub fn increase_borrows(
+        &mut self,
+        account: AccountId,
+        tokens_amount: WBalance,
+    ) -> Balance {
+        let existing_borrows: Balance = self.get_borrows_by_account(account.clone());
+        let increased_borrows: Balance = existing_borrows + Balance::from(tokens_amount);
+
+        self.total_borrows += Balance::from(tokens_amount);
```
Comment (markblaize) [inline contracts/dtoken/src/borrow.rs:28]: @dariacherniak type overflow should be checked on every operation that may cause it

File: contracts/dtoken/src/borrow.rs [hunk 6]
```diff
@@ -1,3 +1,46 @@
...
+
+    #[private]
+    pub fn set_borrows(&mut self, account: AccountId, tokens_amount: WBalance) -> Balance {
+        self.borrows
+            .insert(&account, &Balance::from(tokens_amount));
+        return Balance::from(tokens_amount);
+    }
+
+    pub fn get_borrows_by_account(&self, account: AccountId) -> Balance{
+        if self.borrows.get(&account).is_none(){
```
Comment (markblaize) [inline contracts/dtoken/src/borrow.rs:40]: @dariacherniak this is exceptional situation that we trying to repay for user that has no registered borrows

File: contracts/dtoken/src/repay.rs [hunk 7]
```diff
@@ -1,3 +1,36 @@
 use crate::*;
 
-impl Contract {}
+#[near_bindgen]
+impl Contract {
+
+    pub fn repay(&mut self, token_amount: WBalance) -> PromiseOrValue<U128> {
+        let debt_amount = self.get_borrows_by_account(env::signer_account_id());
+        assert!(Balance::from(token_amount) >= debt_amount*self.get_borrow_rate(), "Amount should be more than borrow * borrow rate");
```
Comment (markblaize) [inline contracts/dtoken/src/repay.rs:8]: @dariacherniak 
```
"repay amount {} is less than existing borrow {}", token_amount, debt_amount * self.get_borrow_rate()
```

File: contracts/controller/src/repay.rs [hunk 8]
```diff
@@ -0,0 +1,38 @@
+use crate::*;
+
+#[near_bindgen]
+impl Contract {
+
+    fn is_repay_allowed(
+        &self,
+        account: AccountId,
+        token_address: AccountId,
+        tokens_amount: WBalance,
```
Comment (markblaize) [inline contracts/controller/src/repay.rs:10]: @dariacherniak tokens_amount -> token_amount (everywhere)

File: contracts/controller/src/repay.rs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    )-> Balance{
+        
+        assert_eq!(
+            self.is_repay_allowed(
+                account_id.clone(),
+                token_address.clone(),
+                tokens_amount.clone(),
+            ),
+            true,
+            "Repay operation is not allowed for account {} token_address {} tokens_amount {}",
```
Comment (markblaize) [inline contracts/controller/src/repay.rs:29]: @dariacherniak 
```
"Repay operation is not allowed for account {} on market {} repay amount {}",
```

File: contracts/dtoken/src/repay.rs [hunk 10]
```diff
@@ -1,3 +1,36 @@
...
+            token_amount,
+            env::current_account_id().clone(),
+            NO_DEPOSIT,
+            self.terra_gas(10),
+        )).into();
+    }
+
+    pub fn controller_repay_borrows_callback(&mut self, amount: WBalance) -> PromiseOrValue<U128> {
+        if !is_promise_success() {
+            log!("Call to decrease user borrows ended incorrect");
```
Comment (markblaize) [inline contracts/dtoken/src/repay.rs:27]: @dariacherniak 
```
log!("failed to decrease user borrows");
```

File: contracts/dtoken/src/supply.rs [hunk 11]
```diff
@@ -16,13 +15,14 @@ impl Contract {
...
-        )))
+        )).into()
     }
-
+ 
     #[allow(dead_code)]
     pub fn supply_balance_of_callback(&mut self, token_amount: WBalance) -> PromiseOrValue<U128> {
 
         if !is_promise_success() {
+            log!("Call to discover dtoken balance ended incorrect");
```
Comment (markblaize) [inline contracts/dtoken/src/supply.rs:25]: @dariacherniak 
```
log!("failed to get {} balance on {}", dtoken, underlyingtoken);
```

File: contracts/dtoken/tests/sim/scenarious.rs [hunk 12]
```diff
@@ -286,7 +306,106 @@ fn scenario_02(){
...
-    let user_balance: u128 = view_balance(&controller, d_user.account_id(), dtoken.account_id());
+    let user_balance: u128 = view_balance(&controller, Supply, d_user.account_id(), dtoken.account_id());
     assert_eq!(user_balance, 0, "Balance should be 0");
 
-}
\ No newline at end of file
+}
+
+#[test]
+fn scenario_03(){
```
Comment (markblaize) [inline contracts/dtoken/tests/sim/scenarious.rs:315]: @dariacherniak 

1) please come up with some long but meaningful scenario name, something like 
```
fn scenario_repay_whole_amount
```
2) once you have checked some functionality you don't need to check it again in other scenarios, so if you have scenario where `utoken.mint` works correctly and if your scenario do not affect `utoken` you don't need to check of result of this function

3) I suppose you have a lot of boilerplate code with initializing state of contracts, in tests such pre-configuration called **Fixture**,

this scenario actually tests only dtoken::repay so all other actions are Fixutre to setup this scenario;
it seems okay to create something like 
scenario_repay_whole_amount_fixutre where to move deploying of accounts and initing balance

**Pull request closed without merging**